Userland/Libraries/LibGUI/Model.h

```diff
@@ -62,13 +62,18 @@ class Model : public RefCounted<Model> {
         MatchFull = 1 << 3,
     };
 
+    struct MatchResult {
+        TriState matched { TriState::Unknown };
```

Comment:
Looking through this, I didn't catch anywhere where any actual model returns TriState::Unknown... Does this actually need to be a tristate? or is a bool for matched good enough? and if a bool is good enough, is `Optional<int>` good enough?

... Looking again, it looks like we return TriState::Unknown from the base class if it's not overridden... Does anyone not override the data_matches function? If we make it pure virtual it could simplify this result thingy... otherwise, never mind :)

Reply:
Unfortunately, there are quite a few Model types that use the default `data_matches` behavior, so I don't think its possible to make it virtual without some kind of bigger restructuring.

After having a closer look at it, it seems that `data_matches` is only used inside `FilteringProxyModel`. It feels like it should be possible to extract the `data_matches` logic from `Model` into a separate `Model::Filter` class that then gets passed in to `FilteringProxyModel`. I'm not sure if I should do that as part of this PR though?

Reply:
yeah a refactor would make sense. I don't think we need to do it now though.